### ISSUE TITLE
[Android] Unknown builtin op: aten::mul

### ISSUE THREAD

**Issue by RuABraun** [pytorch/pytorch#27726]
## 🐛 Bug

When trying to run a single layer model on android I end up getting a SIGABRT:
```
    Unknown builtin op: aten::mul.
    Could not find any similar ops to aten::mul. This op may not exist or may not be currently supported in TorchScript.
    :
    at <string>:3:9
    
    def mul(a : float, b : Tensor) -> Tensor:
      return b * a
             ~~~~~ <--- HERE
    def add(a : float, b : Tensor) -> Tensor:
      return b + a
    'mul' is being compiled since it was called from 'Mod.forward'
    at code/__torch__.py:12:42
        weight = _0.weight
        bias = _0.bias
        _1 = torch.addmm(bias, input, torch.t(weight), beta=1, alpha=1)
                                              ~~~~~~ <--- HERE
        return _1
    Compiled from code /usr/local/lib/python3.6/site-packages/torch/nn/functional.py(1370): linear
    /usr/local/lib/python3.6/site-packages/torch/nn/modules/linear.py(87): forward
    /usr/local/lib/python3.6/site-packages/torch/nn/modules/module.py(525): _slow_forward
    /usr/local/lib/python3.6/site-packages/torch/nn/modules/module.py(539): __call__
    run.py(11): forward
    /usr/
```
Bit confused why it's pointing towards my python pytorch package... 

## To Reproduce

1. Create simple model
```
import torch as to
import torch.nn as nn


class Mod(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(3, 2)

    def forward(self, x): 
        return self.fc(x)


def main():
    model = Mod()
    x = to.rand(1, 3)
    traced_module = to.jit.trace(model, x)

    traced_module.save('traced_model.pt')

main()
```

(1b. Test running locally with C++, this worked for me)

2. Build pytorch master for android using `scripts/build_pytorch_android.sh`. I commented out most of the gradle stuff and just set `GRADLE_PATH` to point to the `gradlew` that my project has. The script does not finish successfully (the gradle part), but creating the static libraries seems to have worked and that's what I care about (not using the java interface).

3. Try running the same code that worked locally on android:
```
    #include "torch/script.h"
    [...]
    torch::jit::script::Module module;
    module = torch::jit::load(modeldir + "traced_model.pt");

    std::vector<torch::jit::IValue> inputs;
    inputs.push_back(torch::ones({1, 3}));
    torch::Tensor output = module.forward(inputs).toTensor();
```

The AndroidStudio CMake has `${PYTORCHDIR}/build_android_${ANDROIDABI}/include` added and I'm linking to libtorch, libc10, libnnpack, libcpuinfo, libeigen_blas and libpytorch_qnnpack.

## Expected behavior

It to run.

## Environment

```
PyTorch version: 1.4.0a0+cddc147
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.14.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GT 1030
GPU 2: GeForce GTX 1080 Ti

Nvidia driver version: 418.87.00
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.3.1
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-9.1/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip3] msgpack-numpy==0.4.3.2
[pip3] numpy==1.16.1
[pip3] torch==1.3.0a0+f433ee1
[pip3] torchvision==0.3.0a0+250bac8
[conda] Could not collect

```

**Comment by xta0** [comment]
Hi @RuABraun, wondering what script you were using to generate the Android build? After 1.3, we have a new building script for mobile - https://pytorch.org/mobile/android/#building-pytorch-android-from-source. Could you try that and see if the error still exists? 

cc @IvanKobzarev @ljk53

**Comment by IvanKobzarev** [comment]
That might be not the answer for your problem:
I tried your model locally using gradle deps 'org.pytorch:pytorch_android:1.3.0' and:
```
    Module module = Module.load(assetFilePath(this, "model_issue_27726.pt"));
    Tensor inputTensor = Tensor.fromBlob(new float[] {1.f, 1.f, 1.f}, new long[] {1, 3});
    final Tensor outputTensor = module.forward(IValue.from(inputTensor)).toTensor();
    final float[] out = outputTensor.getDataAsFloatArray();
```
It works for me on emulator.

I think the problem is with build process of libtorch.
What is the android_abi and android version of device you tried  it to run? (arm64-v8a?)

**Comment by RuABraun** [comment]
Sorry I stupidly copy-pasted my environment from the last issue I created (a week or two ago). The pytorch version is of course different (`1.4.0a0+cddc147`), updated the OP. 

I'll add more info later.

**Comment by gchanan** [comment]
CC @dreiss @jia-kai not sure how you are tracking mobile issues.

**Comment by jia-kai** [comment]
@gchanan Did you accidentally mentioned me? I'm not sure what's going on here

**Comment by xta0** [comment]
> CC @dreiss @jia-kai not sure how you are tracking mobile issues.

cc @ljk53.

**Comment by RuABraun** [comment]
The ABI is `armeabi-v7a`, NDK leve is at 18, android SDK 28.

To use pytorch I just use `include_directories()` to add in `${TORCHPATH}/include`, and then in `target_link_libraries` put in the paths to the static libraries, those are:
```
#${TORCHPATH}/lib/libtorch.a ${TORCHPATH}/lib/libc10.a ${TORCHPATH}/lib/libnnpack.a ${TORCHPATH}/lib/libcpuinfo.a
#${TORCHPATH}/lib/libeigen_blas.a ${TORCHPATH}/lib/libpytorch_qnnpack.a
```
And then use `#include "torch/script.h"`.

I noticed in the `CMakeLists.txt` of pytorch, there is an option:
```
USE_PYTORCH_QNNPACK "Use ATen/QNNPACK (quantized 8-bit operators)"
```
So does using `QNNPACK` mean one uses `ATen`, or is it saying "use ATen **or** QNNPACK"? I'm not quantising my pytorch model actually (maybe that's relevant).

I also do have QNNPACK as a separate dependency in the android app CMakeLists, as another DNN toolkit I'm using uses it. Maybe that's messing things up (doubt it but thought it worth mentioning).

**Comment by ljk53** [comment]
> CC @dreiss @jia-kai not sure how you are tracking mobile issues.

@gchanan to answer your question: we discussed with Edward last week - plan to still rely on master triage oncall to classify mobile issues with the "mobile" label. Mobile team oncall will monitor this label.

**Comment by ljk53** [comment]
> #${TORCHPATH}/lib/libtorch.a ${TORCHPATH}/lib/libc10.a ${TORCHPATH}/lib/libnnpack.a ${TORCHPATH}/lib/libcpuinfo.a
> #${TORCHPATH}/lib/libeigen_blas.a ${TORCHPATH}/lib/libpytorch_qnnpack.a

@RuABraun, looks like you plan to use libtorch c++ library directly - if so you probably should follow this CMake that @IvanKobzarev posted: https://github.com/pytorch/pytorch/blob/master/android/pytorch_android/CMakeLists.txt#L43-L63

Specifically you want to use "-Wl,--whole-archive" before linking "libtorch" as otherwise some operators won't be registered correctly and might cause missing op error at runtime.

**Comment by RuABraun** [comment]
So I modified my CMakeLists like you suggested to use `--whole-archive`:
```
[...]
${QNNPATH}/libqnnpack.a
-Wl,--gc-sections -Wl,--whole-archive ${TORCHPATH}/lib/libtorch.a -Wl,--no-whole-archive ${TORCHPATH}/lib/libc10.a ${TORCHPATH}/lib/libnnpack.a ${TORCHPATH}/lib/libcpuinfo.a ${TORCHPATH}/lib/libclog.a ${TORCHPATH}/lib/libeigen_blas.a ${TORCHPATH}/lib/libpytorch_qnnpack.a
```
And I'm getting a different error now, but it's a very similar one:
```
E/libc++abi: terminating with uncaught exception of type torch::jit::script::ErrorReport: 
    Unknown builtin op: aten::_adaptive_avg_pool2d_backward.
    Could not find any similar ops to aten::_adaptive_avg_pool2d_backward. This op may not exist or may not be currently supported in TorchScript.
```

**Comment by ljk53** [comment]
@RuABraun another thing our java wrapper does which you might miss is to set AutoGradMode before running inference: https://github.com/pytorch/pytorch/blob/master/android/pytorch_android/src/main/cpp/pytorch_jni.cpp#L617
```
torch::autograd::AutoGradMode guard(false);
```

This is because our mobile libtorch library is inference only, which will fail if autograd is enabled.

But I don't see why it complains about missing "_adaptive_avg_pool2d_backward" - your model doesn't call avg_pool2d at all. Are you sure you are using the single layer model? I couldn't repro this error with the code you posted.

**Comment by RuABraun** [comment]
Thanks for the hint about disabling autograd!

I'll make a simple android app with pytorch to make things simpler (I have lots of other C++ dependencies) and make it easier to reproduce the error. Will post tomorrow.

**Comment by RuABraun** [comment]
Oh man, I went through the bother of reproducing the error by creating a simple app, then added `torch::autograd::AutoGradMode guard(false);` and that fixed everything (also in my real app)!

I've uploaded the app code [here](https://github.com/RuABraun/android-with-pytorch). I'm hoping that should make it easy for you to reproduce the `_adaptive_avg_pool2d_backward` error.

**Comment by ljk53** [comment]
> Oh man, I went through the bother of reproducing the error by creating a simple app, then added `torch::autograd::AutoGradMode guard(false);` and that fixed everything (also in my real app)!

Glad to know it helps. Did it fix the "_adaptive_avg_pool2d_backward" error or not?

In case you don't know, we also created a sample android app that you can play with: https://github.com/pytorch/android-demo-app

More information can be found at: https://pytorch.org/mobile/home/

**Comment by RuABraun** [comment]
Yes it fixed it.

The announcement is why I bothered to try and give pytorch on mobile another go (this is my 4th attempt)!
